src/esp/gfx/PbrShader.cpp

```diff
@@ -0,0 +1,481 @@
+    if (lightCount) {
+      bindAttributeLocation(Normal::Location, "vertexNormal");
+    }
+    if (flags_ & (Flag::NormalTexture | Flag::PrecomputedTangent) &&
+        lightCount) {
+      bindAttributeLocation(Tangent4::Location, "vertexTangent");
+    }
+    if (isTextured()) {
+      bindAttributeLocation(TextureCoordinates::Location, "vertexTexCoord");
+    }
+  }
+
+  // set texture binding points in the shader;
+  // see PBR vertex, fragment shader code for details
+  if (flags_ & Flag::BaseColorTexture) {
+    setUniform(uniformLocation("baseColorTexture"), TextureUnit::BaseColor);
+  }
+  if (flags_ & Flag::RoughnessTexture) {
+    setUniform(uniformLocation("roughnessTexture"), TextureUnit::Roughness);
+  }
+  if (flags_ & Flag::MetallicTexture) {
+    setUniform(uniformLocation("metallicTexture"), TextureUnit::Metallic);
+  }
+  if (flags_ & Flag::NormalTexture) {
+    setUniform(uniformLocation("normalTexture"), TextureUnit::Normal);
+  }
+  // emissive texture does NOT need the light!
+  if (flags_ & Flag::EmissiveTexture) {
+    setUniform(uniformLocation("emissiveTexture"), TextureUnit::Emissive);
+  }
+  if ((flags_ & Flag::NoneRoughnessMetallicTexture) ||
+      (flags_ & Flag::OcclusionRoughnessMetallicTexture)) {
+    setUniform(uniformLocation("packedTexture"), TextureUnit::Packed);
+  }
+
+  // cache the uniform locations
+  modelviewMatrixUniform_ = uniformLocation("ModelViewMatrix");
+  normalMatrixUniform_ = uniformLocation("NormalMatrix");
+  mvpMatrixUniform_ = uniformLocation("MVP");
+  if (flags_ & Flag::ObjectId) {
+    objectIdUniform_ = uniformLocation("objectId");
+  }
+  if (flags_ & Flag::TextureTransformation) {
+    textureMatrixUniform_ = uniformLocation("textureMatrix");
+  }
+
+  // materials
+  baseColorUniform_ = uniformLocation("Material.baseColor");
+  roughnessUniform_ = uniformLocation("Material.roughness");
+  metallicUniform_ = uniformLocation("Material.metallic");
+  emissiveColorUniform_ = uniformLocation("Material.emissiveColor");
+
+  // lights
+  if (lightCount_) {
+    lightRangesUniform_ = uniformLocation("LightRanges");
+    lightColorsUniform_ = uniformLocation("LightColors");
+    lightDirectionsUniform_ = uniformLocation("LightDirections");
+
+    if (flags_ & Flag::NormalTexture && flags_ & Flag::NormalTextureScale) {
+      normalTextureScaleUniform_ = uniformLocation("normalTextureScale");
+    }
+  }
+}
+
+// Note: the texture binding points are explicitly specified above.
+// Cannot use "explicit uniform location" directly in shader since
+// it requires GL4.3 (We stick to GL4.1 for MacOS).
+PbrShader& PbrShader::bindBaseColorTexture(Mn::GL::Texture2D* texture) {
+  if ((flags_ & Flag::BaseColorTexture) && lightCount_ && texture) {
+    texture->bind(TextureUnit::BaseColor);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::bindRoughnessTexture(Mn::GL::Texture2D* texture) {
+  if ((flags_ & Flag::RoughnessTexture) && lightCount_ && texture) {
```

Comment:
Are we really not going with references here? What does the coding style guide say about this? Because I remember (mutable) non-nullable references being used quite on a lot of places in Habitat, and using a reference would make it consistent with what Magnum does -- there a pointer is used *only* if it's allowed to be null.

I don't buy the argument that it makes it "easier to use" because MaterialData also use pointers. 

There's an additional concern I have with silent failures -- if the texture is `nullptr` for some reason or the shader doesn't have `RoughnessTexture` enabled for some reason, it will just silently do nothing. In my opinion it neither makes sense to call `bindRoughnessTexture(nullptr)` nor to call `bindRoughnessTexture()` when the shader doesn't have that enabled, so none of that should be allowed:

1. Accept a reference, so if the caller has a pointer for some reason, *the caller* is responsible for dereferencing it. It's not the shader's business to do error handling on behalf of the caller.
2. *Assert* if this function is called and `RoughnessTexture` is not enabled, instead of silently doing a bad render because I don't know, the shader got compiled with packed textures enabled and you're passing separate maps instead. The assert is what Magnum does as well, and from experience I can say it caught quite a lot of mistakes that would take significant time to discover and debug otherwise.

Cc: @aclegg3 @jturner65

Reply:
> I don't buy the argument that it makes it "easier to use" because MaterialData also use pointers.

Yes, for the reason "you do not buy". It is much more convenient to use pointer here. In the render() in PbrDrawable (which is not in this PR), I can use the method chaining to call them, and do not need to care about if the texture exists, or if the shader is initialized with this feature.
```
      .bindBaseColorTexture(materialData_->baseColorTexture)
      .bindRoughnessTexture(materialData_->roughnessTexture)
      .bindMetallicTexture(materialData_->metallicTexture)
      .bindNoneRoughnessMetallicTexture(
          materialData_->noneRoughnessMetallicTexture)
      .bindOcclusionRoughnessMetallicTexture(
          materialData_->occlusionRoughnessMetallicTexture)
      .bindNormalTexture(materialData_->normalTexture)
      .bindEmissiveTexture(materialData_->emissiveTexture)
```
If you change it to the reference (which I tried indeed) you will have a lot of if clause here.

This issue actually reflects different view between us in terms of the robustness of the code.
- I tried to keep the system running, not being interrupted (terminated) due to developers' careless mistakes (NON-fatal errors, e.g., there is actually no roughness texture, but she bound it), and I hided the sanity checks within my APIs to make sure the system will not crash due to these "mistakes". I choose a more tolerant way, and provide easy-to-use APIs (e.g., here you can use method chaining, without any if clauses, and you do not have to remember if you activate corresponding feature or not). "Lazy" developers are going to like it. :)

- You used a more active way to encourage the developer "fail" earlier. As you mentioned, " I can say it caught quite a lot of mistakes that would take significant time to discover and debug otherwise." which I agree. 

In short, the difference is that you leave such overhead (sanity check, dereferencing, assert etc.) to the developer, but I do not. No right or wrong here. Just different style.

I just explain to let you know my view. And I do not think one is more appealing than the other one.

I can change to the way you suggested. I am OK with it.

Reply:
Funnily enough right now I'm fighting `glslang` which is built with this "fail silently" approach and it costs me an immense amount of time because a lot of the time it just does nothing and i don't know *why* :laughing: 

Yep, It's a matter of perspective. Could we get opinions from others here? I don't want to push my view if I would the only one who wants it that way.

Reply:
I changed them to use the reference here, and also added assertions within the API to make it fail immediately in illegal usage cases. It is worth if "it caught quite a lot of mistakes that would take significant time to discover and debug otherwise." We can continue the conversation on this topic in the slack channel.

This PR is out for quite a long time, and I hope there is no more further delays.

Reply:
@eundersander : Here is the discussion on the design philosophy.